scripts/generate-cli-completions.py

```
@@ -0,0 +1,419 @@
+            )
+        return text
+    # Canonical default: parse the heredoc straight from source.
+    return extract_usage_heredoc(Path(args.source))
+
+
+def parse_registry(source_path: Path) -> list[str]:
+    """Extract command names from the `topLevelCommandNames` Set literal."""
+    text = source_path.read_text(encoding="utf-8")
+    start = text.find("topLevelCommandNames: Set<String> = [")
+    if start == -1:
+        raise SystemExit(f"topLevelCommandNames not found in {source_path}")
```

Ruff (0.15.18)

[warning] Avoid specifying long messages outside the exception class

(TRY003)